gpflow/models/model.py

```diff
@@ -25,7 +25,12 @@
 from ..utilities import to_default_float
 from .training_mixins import InputData, RegressionData
 
-MeanAndVariance = Tuple[tf.Tensor, tf.Tensor]
+
+class MeanAndVariance(NamedTuple):
+    """ NamedTuple to access mean- and variance-function separately """
+
+    mean: tf.Tensor
+    variance: tf.Tensor
```

Comment:
More of a general question at other maintainers (@vdutor @markvdw @awav) & anyone else ...: what should the name of this field be? In code we often abbreviate the variance to "var", but also it sometimes represents the covariance... or maybe that should be a different type? MeanAndVariance and MeanAndCov (and return a different one depending on full_cov etc)?

Reply:
i like the idea of the two types. Haven't thought about it in detail

Reply:
do you always know for sure if you've got the cov or var?

Reply:
If you pass in full_cov=False and full_output_cov=False, you get the marginals back. If one of full_cov or full_output_cov is True, you get the covariance over inputs or outputs, respectively. If both are True, you should get the N P x N P covariance matrix (though this combination isn't actually implemented in several cases, I believe). So the output type is solely determined by the full_cov and full_output_cov arguments.

Reply:
btw you can use `typing.overload` and `typing.Literal` to give more information than just `MeanAndVariance | MeanAndCov`, if you wanted to do that. Probably not sth for this PR though

Reply:
> btw you can use typing.overload and typing.Literal to give more information than just MeanAndVariance | MeanAndCov, if you wanted to do that

Alright! I'll look into it the next week :D

Reply:
@joelberkeley , do you suggest adding Literals to the class-attributes-type and overloading the "variance" attribute by "covariance" (and the class-name too)? :)

Reply:
i'd suggest leaving it for this PR. It's not vital, it's fiddly and tbh I've not used `Literal` yet so I'm kind of guessing how it works. Plain overloads would let you do sth like this (I've omitted some arguments)
```python
@overload
def predict_f(self, ...) -> MeanAndVariance: ...

def predict_f(self, ..., full_cov: bool = False) -> Union[MeanAndCovariance, MeanAndVariance]:
    ...
```
for literals you could add
```python
@overload
def predict_f(self, ..., full_cov: Literal[True]) -> MeanAndCovariance: ...

@overload
def predict_f(self, ..., full_cov: Literal[False]) -> MeanAndVariance: ...
```

Reply:
Ah. I thought of using it in the definition of `MeanAndVariance | MeanAndCov`. I didn't consider using `Literal` and `overload` in `predict_f`.  That means that the user would see different return types of `predict_f` depending on full_cov, etc. Sounds pretty nice!

Regardless of this PR, I would play a bit with this idea :D

Reply:
I've created a small example, which should work as wanted:
![on_hover](https://user-images.githubusercontent.com/45163503/115148462-010f0c00-a060-11eb-9e8c-2b582cd5fda8.png)
![on_space](https://user-images.githubusercontent.com/45163503/115148466-0409fc80-a060-11eb-8c61-e35b3f3eb54a.png)

But I' not comfortable with the if-else statement :/

```python
from typing import Literal, NamedTuple, overload

class MeanAndVariance(NamedTuple):
    mean: int
    variance: int

class MeanAndCovariance(NamedTuple):
    mean: int
    covariance: int

@overload
def predict_f(auto_cov: Literal[False]) -> MeanAndVariance:
    ...

@overload
def predict_f(auto_cov: Literal[True]) -> MeanAndCovariance:
    ...

def predict_f(auto_cov: bool = False) -> (MeanAndVariance | MeanAndCovariance):
    # calculations
    return  MeanAndCovariance(1, 2) if auto_cov else MeanAndVariance(1, 2)
```

Reply:
I don't think there's a good way around the if-else, and I think it's better to be explicit than the ambiguity of having to remember whether it's a [N, Q] or [Q, N, N] tensor ...:) I'd be happy with this.

Reply:
As it pointed out, [typing.Literal](https://docs.python.org/3/library/typing.html#typing.Literal) is only available from Python 3.8 and up :D

Reply:
@antonykamp the [typing_extensions module](https://github.com/python/typing/blob/master/typing_extensions/README.rst) provides backports for older versions of Python, it does seem to include Literal. :)

Reply:
I added this pattern with `overload` and `Tuple` to the abstract method  `predict_f` in `gpflow/models/models.py`. Should the overloads of `predict_f` be marked as _abstract_ too? In any case I see no chance to test the Ellipsis operator of each overloaded funciton :/

Also, I wanted to ask if the parameters of the model constructor should be listed in the parametrization?